### ISSUE TITLE
Lua coroutine API may cause C stack overflow

### ISSUE THREAD

**Issue by neoxic** [LuaJIT/LuaJIT#619]
LuaJIT latest v2.1 HEAD

test.lua
---
```Lua
local function f()
    coroutine.wrap(f)()
end
f()
```

```
$ luajit-2.1.0-beta3 test.lua
Segmentation fault
```
LuaJIT 2.0.5 is affected too.

**Comment by neoxic** [comment]
Might be related to #618

**Comment by MikePall** [comment]
This is a C stack overflow and unrelated to #618.

Due to a general design issue with the Lua coroutine API, any invocation of a coroutine always takes up C stack space, too. Plain Lua has various ineffective mitigations in different versions, but it's always possible to construct a testcase that blows up the C stack (see the recent Lua mailing list threads about 5.4). Mainly, because there's no portable way to detect when the C stack is going to overflow (before it does so with a SEGV).

In other words: this is unresolvable in general without breaking compatibility with the Lua API.
